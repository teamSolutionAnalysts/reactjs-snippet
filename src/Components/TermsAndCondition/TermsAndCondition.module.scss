@import "../../../public/static/css/variable.scss";

.termsAndConditionContainer {
    input[type="text"].formControl {
        display: flex;
        width: 100%;
        padding: 10px 20px;
        font-size: 1.188em;
        line-height: 1.5;
        color: #000;
        background-color: #f2efea;
        background-clip: padding-box;
        border: 1px solid #f2efea;
        border-radius: 30px;
        height: 60px;
        font-weight: 500;
    }

    .formControl,
    input[type="text"].formControl {
        display: flex;
        width: 100%;
        padding: 10px 20px;
        font-size: 1.188em;
        line-height: 1.5;
        color: #000;
        background-color: #f2efea;
        background-clip: padding-box;
        border: 1px solid #f2efea;
        border-radius: 30px;
        height: 60px;
        font-weight: 500;
    }

    .formControl:focus,
    input[type="text"].formControl:focus {
        border-color: rgba(0, 0, 0, 0.2);
        background-color: #fff;
        outline: 0;
        box-shadow: none;
    }

    .formControl::-webkit-input-placeholder {
        color: #495057;
        opacity: 1;
    }

    .formControl::-moz-placeholder {
        color: #495057;
        opacity: 1;
    }

    .formControl:-ms-input-placeholder {
        color: #495057;
        opacity: 1;
    }

    .formControl::-ms-input-placeholder {
        color: #495057;
        opacity: 1;
    }

    .formControl::placeholder {
        color: #495057;
        opacity: 1;
    }

    .customDrop {
        padding: 15px 20px;
    }

    .buttonColEnd {
        display: flex;
        justify-content: flex-end;
        margin-top: 1.5em;
    }

    .customBtn {
        height: 40px;
        border: none;
        padding: 5px 20px;
        max-width: 130px;
        width: 100%;

        @media (max-width: 767px) {
            font-size: $font-size-18;
            max-width: 100px;
        }
    }

    .blackBtn {
        background-color: $black;
        color: $white;
        margin-right: 1em;
    }

    .yellowBtn {
        color: $black;
        background-color: $nav-bg;
        font-weight: 700;

        &:hover {
            text-decoration: none;
            background-color: #000;
            color: #fff;
            border-color: #000;
        }
    }
    .smallLoader {
        color: #000000;
        position: absolute;
        left: 0px;
        right: 0px;
        cursor: default;
    }
    .checkboxContainer {
        display: flex;
        align-items: center;
        .checkboxIconContainer {
            display: flex;
            flex: 1;
            background-color: $nav-bg;
            align-self: stretch;
        }
    }
    .checkboxLabelText {
        margin-left: 10px;
    }
}

.arabic-body .termsAndConditionContainer {
    text-align: right;
}

.arabic-body .checkboxLabelText {
    margin-left: 0px;
    margin-right: 10px;
}

.disable {
    opacity: 0.6;
}