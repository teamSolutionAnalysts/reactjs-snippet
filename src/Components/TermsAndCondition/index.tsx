/* eslint-disable react-hooks/exhaustive-deps */
// #region Global Imports
import React, { useEffect, useState } from "react";
import { NextPage } from "next";
import { Formik } from "formik";
import { WithTranslation } from "next-i18next";
import { Button } from "antd";
import { withTranslation } from "@Server/i18n";
import { STLCol, STLRow, CustomCheckbox, Loader } from "@Components/Common";
import { CheckOutlined } from "@ant-design/icons";
import { useSelector, useDispatch } from "react-redux";
import { PLAN_DETAILS } from "@Utils/enums";
import { NotificationWithIcon } from "@Utils/helper";
import CheckoutService from "@Services/API/Checkout";
import CheckoutActions from "@Redux/Actions/CheckoutActions";
import CartService from "@Services/API/Cart";
import styles from "./TermsAndCondition.module.scss";

interface TermsAndConditionProps extends WithTranslation {
    successMainDetailsHandle?: any;
    termsAndConditionsData?: any;
    getUserDetailsForBillingHandler?: any;
}

const TermsAndCondition: NextPage<TermsAndConditionProps> = ({
    i18n,
    successMainDetailsHandle,
    t,
    termsAndConditionsData,
}: any) => {
    const dispatch = useDispatch();
    const planState = useSelector((state: any) => state.plan);
    const cartState = useSelector((state: any) => state.cart);
    const { relatedPartyId, shoppingCartId } = cartState;
    const { paymentMode } = planState;
    const [loader, setLoader] = useState(false);
    const [termsCheck, setTermsCheck] = useState<boolean | null>(null);
    const commonState = useSelector((state: any) => state.common);
    const { checkoutData } = commonState;

    useEffect(() => {
        if (checkoutData) {
            setTermsCheck(checkoutData?.termsAndCondition || false);
        } else {
            setTermsCheck(false);
        }
    }, []);

    const callAcceptTNC = async () => {
        setLoader(true);
        try {
            await CartService.acceptTNC({ termsAndCondition: true });
            return true;
        } catch (error) {
            const err: any = error;
            NotificationWithIcon("error", err?.data?.responseMessage);
            return false;
        }
    };

    if (termsCheck === null) return null;

    return (
        <div className={i18n.language === "ar" ? `${styles["arabic-body"]}` : ""}>
            {typeof termsAndConditionsData?.body !== "undefined" ? (
                <div className={styles.termsAndConditionContainer}>
                    <div
                        data-testid="termsConditionData"
                        // eslint-disable-next-line react/no-danger
                        dangerouslySetInnerHTML={{ __html: termsAndConditionsData?.body }}
                    />
                    <Formik
                        initialValues={{
                            termsCheck,
                        }}
                        enableReinitialize
                        onSubmit={() => {
                            callAcceptTNC()
                                .then(() => {
                                    if (paymentMode === PLAN_DETAILS.Prepaid) {
                                        const data = {
                                            relatedPartyId,
                                            shoppingCartId,
                                        };
                                        CheckoutService.billingAccountPrepaidUpdate(data)
                                            .then((res: any) => {
                                                dispatch(
                                                    CheckoutActions.SetbaNumber({
                                                        baNumber: res?.result?.baNumber,
                                                    })
                                                );
                                                dispatch(
                                                    CheckoutActions.SetsaNumber({
                                                        saNumber: res?.data?.saNumber,
                                                    })
                                                );
                                                successMainDetailsHandle(["orderSummery"], "termsAndConditions");
                                                // getUserDetailsForBillingHandler().then(() => {
                                                //     setLoader(false);
                                                // });
                                            })
                                            .catch((err: any) => {
                                                setLoader(false);
                                                NotificationWithIcon(
                                                    "error",
                                                    err?.data?.error || "Something went wrong"
                                                );
                                            });
                                    } else {
                                        successMainDetailsHandle(["orderSummery"], "termsAndConditions");
                                    }
                                })
                                .catch(err => {
                                    console.log("Error", err);
                                });
                        }}
                        render={({ values, handleSubmit, handleChange, setFieldValue }) => {
                            return (
                                <>
                                    <STLRow gutter={[16, 16]}>
                                        <STLCol lg={12} md={12}>
                                            <div className={styles.checkboxContainer}>
                                                <CustomCheckbox
                                                    checked={values?.termsCheck}
                                                    icon={
                                                        <div className={styles.checkboxIconContainer}>
                                                            <CheckOutlined color="#fff" />
                                                        </div>
                                                    }
                                                    borderColor="#8c8c8c80"
                                                    containerStyle={{ marginBottom: 0 }}
                                                    borderRadius={2}
                                                    size={20}
                                                    setFieldValue={setFieldValue}
                                                    filedName="termsCheck"
                                                    onChange={handleChange}
                                                />

                                                <span className={styles.checkboxLabelText}>
                                                    {termsAndConditionsData?.nothing}
                                                </span>
                                            </div>
                                        </STLCol>
                                    </STLRow>
                                    <STLRow gutter={[16, 16]}>
                                        <STLCol span={24} className={styles.buttonColEnd}>
                                            <Button
                                                type="primary"
                                                shape="round"
                                                className={`${styles.customBtn} ${
                                                    !values?.termsCheck || loader
                                                        ? styles.disableButton
                                                        : styles.yellowBtn
                                                }`}
                                                disabled={!values?.termsCheck || loader}
                                                onClick={(e: any) => handleSubmit(e)}
                                                data-testid="termsButton"
                                            >
                                                {loader && <Loader divClassName={styles.smallLoader} />}
                                                {t("common:SAVE")}
                                            </Button>
                                        </STLCol>
                                    </STLRow>
                                </>
                            );
                        }}
                    />
                </div>
            ) : null}
        </div>
    );
};

export default withTranslation(["common", "termsAndConditions"])(TermsAndCondition);
