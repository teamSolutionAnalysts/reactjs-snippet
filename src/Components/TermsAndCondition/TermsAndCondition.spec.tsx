import React from "react";

import { setupServer } from "msw/node";
import { act, fireEvent, render } from "@Test/utils";
import { Provider } from "react-redux";
import * as redux from "react-redux";
import { store } from "@Redux";

import "@testing-library/jest-dom/extend-expect";

import TermsAndCondition from ".";

const server = setupServer();

beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

const data = {
    body: "This is terms and condition for project",
};

describe("TermsAndCondition render Page", () => {
    const useSelectorMock = jest.spyOn(redux, "useSelector");

    beforeEach(() => {
        useSelectorMock.mockClear();
    });

    it("TermsAndCondition snapshot", () => {
        useSelectorMock.mockReturnValue({
            paymentMode: "prepaid",
        });
        const { container } = render(
            <Provider store={store}>
                <TermsAndCondition termsAndConditionsData={data} successMainDetailsHandle={() => jest.fn()} />
            </Provider>
        );
        expect(container).toMatchSnapshot();
    });

    it("TermsAndCondition text", () => {
        useSelectorMock.mockReturnValue({
            paymentMode: "prepaid",
        });
        const { getByTestId } = render(
            <Provider store={store}>
                <TermsAndCondition termsAndConditionsData={data} successMainDetailsHandle={() => jest.fn()} />
            </Provider>
        );
        const termsText = getByTestId("termsConditionData");

        expect(data.body).not.toBe(undefined);
        act(() => {
            expect(termsText).toHaveTextContent(data.body);
        });
    });

    it("renders a submit button", () => {
        useSelectorMock.mockReturnValue({
            paymentMode: "prepaid",
        });
        const { getByTestId } = render(
            <Provider store={store}>
                <TermsAndCondition termsAndConditionsData={data} successMainDetailsHandle={() => jest.fn()} />
            </Provider>
        );
        expect(getByTestId("termsButton")).toBeInTheDocument();
    });

    it("allows the user to allow terms and condition and save", async () => {
        useSelectorMock.mockReturnValue({
            paymentMode: "prepaid",
        });
        const { getByTestId } = render(
            <Provider store={store}>
                <TermsAndCondition termsAndConditionsData={data} successMainDetailsHandle={() => jest.fn()} />
            </Provider>
        );

        expect(getByTestId("termsButton")).toHaveAttribute("disabled");

        fireEvent.click(getByTestId("checkbox"));

        await act(async () => {
            expect(getByTestId("termsButton")).not.toHaveAttribute("disabled");
        });
    });
});
