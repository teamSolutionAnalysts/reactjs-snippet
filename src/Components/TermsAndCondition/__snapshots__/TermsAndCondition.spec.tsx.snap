// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TermsAndCondition render Page TermsAndCondition snapshot 1`] = `
<div>
  <div
    class=""
  >
    <div>
      <div
        data-testid="termsConditionData"
      >
        This is terms and condition for project
      </div>
      <div
        class="ant-row"
        style="margin: -8px -8px -8px -8px;"
      >
        <div
          class="ant-col ant-col-md-12 ant-col-lg-12"
          style="padding: 8px 8px 8px 8px;"
        >
          <div>
            <p
              class=""
              style="margin-bottom: 0px; display: flex; align-items: center;"
            >
              <span>
                <div
                  class=""
                  style="height: 20px; width: 20px; border-width: 2px; border-color: #8c8c8c80; border-style: solid; border-radius: 2px; display: flex; align-items: center; justify-content: center;"
                >
                  <input
                    data-testid="checkbox"
                    hidden=""
                    name=""
                    type="checkbox"
                    value=""
                  />
                </div>
              </span>
            </p>
            <span />
          </div>
        </div>
      </div>
      <div
        class="ant-row"
        style="margin: -8px -8px -8px -8px;"
      >
        <div
          class="ant-col ant-col-24"
          style="padding: 8px 8px 8px 8px;"
        >
          <button
            class="ant-btn ant-btn-primary ant-btn-round undefined undefined"
            data-testid="termsButton"
            disabled=""
            type="button"
          >
            <span>
              common:SAVE
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
